tests/csapi/apidoc_room_event_test.go

```diff
@@ -0,0 +1,206 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+)
+
+func stringSliceContains(slice []string, s string) bool {
+	for _, value := range slice {
+		if value == s {
+			return true
+		}
+	}
+	return false
+}
+
+// sytest: Events come down the correct room
+func TestEventsInCorrectRoom(t *testing.T) {
+	deployment := Deploy(t, b.BlueprintAlice)
+	defer deployment.Destroy(t)
+
+	alice := deployment.Client(t, "hs1", "@alice:hs1")
+
+	const (
+		roomAmount     = 10
+		goConcurrency  = 4
+		channelTimeout = 10 * time.Second
+	)
+
+	// Create all rooms
+
+	var doneRooms = make(chan string, roomAmount)
+	var todoRooms = make(chan bool, goConcurrency)
+
+	for i := 0; i < goConcurrency; i++ {
+		go func() {
+			for {
+				_, ok := <-todoRooms
+
+				if !ok {
+					return
+				}
+
+				roomId := alice.CreateRoom(t, map[string]interface{}{
+					"preset": "public_chat",
+				})
+
+				doneRooms <- roomId
+			}
+		}()
+	}
+
+	for i := 0; i < roomAmount; i++ {
+		todoRooms <- true
+	}
+
+	close(todoRooms)
+
+	var rooms []string
+
+	for i := 0; i < roomAmount; i++ {
+		select {
+		case newRoom := <-doneRooms:
+			rooms = append(rooms, newRoom)
+		case <-time.After(channelTimeout):
+			t.Fatalf("Room creation timed out, got to %d rooms (from %d)", i, roomAmount)
+		}
+	}
+
+	// Send events to all rooms with corresponding roomID
+
+	// get current next_batch
+	_, since := alice.MustSync(t, client.SyncReq{TimeoutMillis: "0"})
```

Comment:
There are no guarantees that this token will be _after_ all the rooms have been created. You likely need a sentinel event here to know for sure that the previous requests have been processed.

Reply:
I'm assuming that a sync request like this, after a response from the room creation endpoint, will include those created rooms, is this a bad assumption?

https://github.com/matrix-org/matrix-spec/issues/938 is related to this issue, btw.

Reply:
In this current form it is probably fine to sync now, as you've waited for all /createRoom calls to 200 OK. To be sure, you could just hit /sync until you see all `roomAmount` rooms.